components/browser/engine-gecko/src/test/java/mozilla/components/browser/engine/gecko/GeckoEngineSessionTest.kt

```diff
@@ -144,13 +143,51 @@ class GeckoEngineSessionTest {
     fun testGoForward() {
         val engineSession = GeckoEngineSession(mock(GeckoRuntime::class.java))
         var eventReceived = false
-        engineSession.geckoSession.eventDispatcher.registerUiThreadListener(object : BundleEventListener {
-            override fun handleMessage(event: String?, message: GeckoBundle?, callback: EventCallback?) {
-                eventReceived = true
-            }
-        }, "GeckoView:GoForward")
+        engineSession.geckoSession.eventDispatcher.registerUiThreadListener(
+                BundleEventListener { _, _, _ -> eventReceived = true },
+                "GeckoView:GoForward"
+        )
 
         engineSession.goForward()
         assertTrue(eventReceived)
     }
+
+    @Test
+    fun testSaveState() {
+        val engineSession = GeckoEngineSession(mock(GeckoRuntime::class.java))
+        engineSession.geckoSession = mock(GeckoSession::class.java)
+        val currentState = GeckoSession.SessionState("")
+        val stateMap = mapOf(GeckoEngineSession.GECKO_STATE_KEY to currentState.toString())
+
+        `when`(engineSession.geckoSession.saveState(any())).thenAnswer(
+                { inv -> (inv.arguments[0] as GeckoResponse<GeckoSession.SessionState>).respond(currentState) })
+
+        assertEquals(stateMap, engineSession.saveState())
+    }
+
+    @Test
+    fun testSaveStateThrowsExceptionOnNullResult() {
+        val engineSession = GeckoEngineSession(mock(GeckoRuntime::class.java))
+        engineSession.geckoSession = mock(GeckoSession::class.java)
+        `when`(engineSession.geckoSession.saveState(any())).thenAnswer(
+                { inv -> (inv.arguments[0] as GeckoResponse<GeckoSession.SessionState>).respond(null) })
+
+        try {
+            engineSession.saveState()
+            fail("Expected GeckoEngineException")
+        } catch (e: GeckoEngineException) { }
+    }
```

Comment:
\[ktlint\] WARNING: \[keyword\-spacing\] Unexpected newline before "catch" in column 9